File: program/program/move-all-zeroes-to-end-of-array/MoveAllZeroesToEndOfArray.ts
function moveZeroesToEnd(arr: number[]): number[] {
    let nonZeroIndex = 0;

    // Move all non-zero elements to the beginning of the array
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] !== 0) {
            arr[nonZeroIndex++] = arr[i];
        }
    }

    // Fill remaining positions with zeroes
    for (let i = nonZeroIndex; i < arr.length; i++) {
        arr[i] = 0;
    }

    return arr;
}

// Test cases
console.log(moveZeroesToEnd([1, 2, 0, 4, 3, 0, 5, 0])); // Output: [1, 2, 4, 3, 5, 0, 0, 0]
console.log(moveZeroesToEnd([1, 2, 0, 0, 0, 3, 6]));   // Output: [1, 2, 3, 6, 0, 0, 0]

File: program/program/move-all-zeroes-to-end-of-array/move_all_zeroes_to_end_of_array.dart
// Dart program to move all zeroes to end of array

void moveZeroes(List<int> arr) {
  int n = arr.length;

  // Traverse the array and move non-zero elements to the beginning
  int count = 0;
  for (int i = 0; i < n; i++) {
    if (arr[i] != 0) {
      arr[count++] = arr[i];
    }
  }

  // Fill the remaining positions with zeroes
  while (count < n) {
    arr[count++] = 0;
  }
}

void main() {
  // Test cases
  List<int> arr1 = [1, 2, 0, 4, 3, 0, 5, 0];
  moveZeroes(arr1);
  print('Output for arr1: $arr1'); // Output : [1, 2, 4, 3, 5, 0, 0, 0]

  List<int> arr2 = [1, 2, 0, 0, 0, 3, 6];
  moveZeroes(arr2);
  print('Output for arr2: $arr2'); // Output : [1, 2, 3, 6, 0, 0, 0]
}
 